### ISSUE TITLE
Copy error

### ISSUE THREAD

**Issue by Tazintosh** [cloudfs/ftp-cloudfs#44]
Hi Juan,

Since this morning, I'm trying to backup an audio file and always get a copy error.
The file is 523,50 MB, my split is set to 512 MB and the transfert always fail. After checking the file size online, it's always 512 MB.
I've been able to upload dozen of similar audio files, about 509 MB.
Any idea?

	[I 15-02-10 09:38:40] 192.168.1.13:60509-[hubic] STOR /default/myAudioTrack.aiff 125
	[75570] 2015-02-10 09:38:40,343 - INFO - 192.168.1.13:60509-[hubic] STOR /default/myAudioTrack.aiff 125
	[75570] 2015-02-10 09:41:46,571 - INFO - Starting new HTTPS connection (1): myServer.com
	[76175] 2015-02-10 09:41:46,574 - INFO - Starting new HTTPS connection (1): myServer.com
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	[76175] 2015-02-10 09:41:49,457 - INFO - REQ: curl -i https://myServer.com/auth/v1.0/ -X GET
	[76175] 2015-02-10 09:41:49,457 - INFO - RESP STATUS: 500 Internal Server Error
	[76175] 2015-02-10 09:41:49,457 - INFO - RESP HEADERS: [('content-length', '11'), ('x-powered-by', 'PHP/5.5.14'), ('expires', 'Mon, 1 Jan 2000 00:00:00 GMT'), ('keep-alive', 'timeout=15, max=100'), ('ms-author-via', 'DAV'), ('server', 'Apache'), ('last-modified', 'Tue, 10 Feb 2015 08:41:49 GMT'), ('connection', 'Keep-Alive'), ('pragma', 'no-cache'), ('cache-control', 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0'), ('date', 'Tue, 10 Feb 2015 08:41:46 GMT'), ('content-type', 'text/html')]
	[76175] 2015-02-10 09:41:49,458 - INFO - RESP BODY: ERROR 178: 
	Process Process-2062:1:
	Traceback (most recent call last):
	  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    	self.run()
	  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 114, in run
	    self._target(*self._args, **self._kwargs)
	  File "/Library/Python/2.7/site-packages/ftpcloudfs/fs.py", line 206, in copy_task
    	url, token = conn.get_auth()
	  File "/Library/Python/2.7/site-packages/ftpcloudfs/fs.py", line 95, in get_auth
	    return super(ProxyConnection, self).get_auth()
	  File "/Library/Python/2.7/site-packages/python_swiftclient-2.3.1-py2.7.egg/swiftclient/client.py", line 1209, in get_auth
	    insecure=self.insecure)
	  File "/Library/Python/2.7/site-packages/python_swiftclient-2.3.1-py2.7.egg/swiftclient/client.py", line 376, in get_auth
	    insecure=insecure)
	  File "/Library/Python/2.7/site-packages/python_swiftclient-2.3.1-py2.7.egg/swiftclient/client.py", line 270, in get_auth_1_0
	    http_status=resp.status, http_reason=resp.reason)
	ClientException: Auth GET failed: https://myServer.com/auth/v1.0/ 500 Internal Server Error
	[75570] 2015-02-10 09:41:50,821 - INFO - Starting new HTTPS connection (1): lb1.hubic.ovh.net
	[I 15-02-10 09:41:51] 192.168.1.13:60509-[hubic] Data connection error ([Errno 5] Failed to store the file)
	[75570] 2015-02-10 09:41:51,438 - INFO - 192.168.1.13:60509-[hubic] Data connection error ([Errno 5] Failed to store the file)
	[I 15-02-10 09:41:51] 192.168.1.13:60509-[hubic] FTP session closed (disconnect).
	[75570] 2015-02-10 09:41:51,440 - INFO - 192.168.1.13:60509-[hubic] FTP session closed (disconnect).
	[I 15-02-10 09:42:00] 192.168.1.13:61971-[] FTP session opened (connect)
	[7442] 2015-02-10 09:42:00,381 - INFO - 192.168.1.13:61971-[] FTP session opened (connect)
	[I 15-02-10 09:42:00] 192.168.1.13:61971-[] FTP session closed (disconnect).

**Comment by reidrac** [comment]
The auth seems to be returning an error: `ClientException: Auth GET failed: https://myServer.com/auth/v1.0/ 500 Internal Server Error`

That could explain why it didn't store the file. Can you verify you can upload smaller files at the moment? There's any kind of quotas check in your storage server?

**Comment by Tazintosh** [comment]
Thanks for the answer Juan,

I've seen this error yes, but I can't see any relation to the actual problem.
• I've no quotas (well, it's 10TB, I'm not yet at this point)
• Privileges, owner and group of the file are the same as all my other ones

• I can successfully upload smaller and much bigger files. But wait, now, I've been able to put the file online… I'm trying for more than three hours here (swear), I've been able to upload any other files but this one. I tested again before answering your question and this time it worked… Idea?

**Comment by reidrac** [comment]
If the authentication fails, you don't get a token and the file upload won't work. The way the ftp server works may looks weird, but we don't verify the auth until we actually need a token not when the user logs in.

It doesn't look to me like the name has anything funny on it. Do you have any other file with a similar name in the same folder?

Other than that, the auth error is important. I would be inclined to think your storage provide has problems.

**Comment by Tazintosh** [comment]
Let's poke @VincentCasse to this discussion :)

**Comment by VincentCasse** [comment]
The auth url looks strange ! https://myServer.com/auth/v1.0/.
I think your app try to relog on keystone. But hubiC doesn't provide Keystone component. You had to use the hubic API to authenticate and get your token for swift.

@Tazintosh what do you use to login into hubic and cloudfs?

**Comment by Tazintosh** [comment]
Hey Vincent, thanks for your visit here.
I'm login with hubic2swiftgate from @oderwat
<https://github.com/oderwat/hubic2swiftgate>

**Comment by Tazintosh** [comment]
Hey guys,
I really have an issue here.

Don't know if it's hubiC side or ftp-clouds (or even my own side), but something is really going wrong.
I've a fail 9 times of 10 (I swear). If even if I've fiber optic here, I cannot not just waste my time re-uploading again, and again, and again, dozen of GB. (Because I can event not "continue" a failed transfert).
I've been able to test with other files, it's always the same story, the split cause a fail.

Would really appreciate any help.

**Comment by reidrac** [comment]
With the information we have, the auth error is important. If the auth service reports an error 500 I'm not sure if that's a problem related to the ftp-cloudfs.

**Comment by reidrac** [comment]
Default token TTL in swift is 86400 seconds. Does that ring a bell?

Anyway, it doesn't sound right that you get a 500 error even if the token had expired.

Looking at hubic2swiftgate the error is triggered at: https://github.com/oderwat/hubic2swiftgate/blob/master/www/simple.php#L311

You can get to that point because you don't have a token or if the token has expired. Apparently the call to "refresh" the token is failing as api.hubic.com is not returning a 200 status.

hubic2swiftgate could do a better job reporting the issue (like including the body of api.hubic.com response, perhaps?).

Now, without not knowing anything else; can you confirm that the box running hubic2swiftgate has the right date/time? I know is a stab in the dark, but without knowing what is the problem with api.hubic.com, that's the best I got!

As I said, this doesn't look like an ftp-cloudfs issue.

Please let us posted about this! :)

**Comment by Tazintosh** [comment]
Hi Juan and thanks for trying out to help me here.
The timeout should not be the problem, uploading the 512 first MB (split setting) is a matter of 30s when my 200MB/s line is fully triggered.
The date/time is perfectly set too.

The only thing I notice is that when the upload reach the split point (512MB), I can clearly see that the connection dropdown for quite a few second before starting again to upload… or failing.

Here is a typical example of failing (I've just done it, file is above 600MB, speed was only ≈30MB/s).
It started to upload without problem (time is 12:32:28), then reach 512MB (time is 12:35:08), then transfert cut for some seconds, then finish by fail (probably time 12:36:24).

	[6543] 2015-03-04 12:32:28,546 - INFO - Starting new HTTPS connection (1): lb1.hubic.ovh.net
	[I 15-03-04 12:32:28] 192.168.1.13:49821-[hubic] STOR /default/myFile.ext 125
	[6543] 2015-03-04 12:32:28,610 - INFO - 192.168.1.13:49821-[hubic] STOR /default/myFile.ext 125
	[6543] 2015-03-04 12:35:08,927 - INFO - Starting new HTTPS connection (1): myServer.com
	[6755] 2015-03-04 12:35:08,929 - INFO - Starting new HTTPS connection (1): myServer.com
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	[6543] 2015-03-04 12:35:10,282 - INFO - Starting new HTTPS connection (1): lb1.hubic.ovh.net
	[6755] 2015-03-04 12:35:10,297 - INFO - Starting new HTTPS connection (1): lb1.hubic.ovh.net
	[I 15-03-04 12:36:24] 192.168.1.13:50041-[] FTP session opened (connect)
	[1399] 2015-03-04 12:36:24,328 - INFO - 192.168.1.13:50041-[] FTP session opened (connect)
	[I 15-03-04 12:36:24] 192.168.1.13:50041-[] FTP session closed (disconnect).
	[1399] 2015-03-04 12:36:24,329 - INFO - 192.168.1.13:50041-[] FTP session closed (disconnect).
	[6777] 2015-03-04 12:36:24,365 - INFO - Starting new HTTPS connection (1): myServer.com
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	[I 15-03-04 12:36:25] 192.168.1.13:50041-[hubic] Authentication validated for user hubic
	[6777] 2015-03-04 12:36:25,670 - INFO - 192.168.1.13:50041-[hubic] Authentication validated for user hubic
	[I 15-03-04 12:36:25] 192.168.1.13:50041-[hubic] USER 'hubic' logged in.
	[6777] 2015-03-04 12:36:25,671 - INFO - 192.168.1.13:50041-[hubic] USER 'hubic' logged in.
	[I 15-03-04 12:36:25] 192.168.1.13:50042-[] FTP session opened (connect)
	[1399] 2015-03-04 12:36:25,675 - INFO - 192.168.1.13:50042-[] FTP session opened (connect)
	[6777] 2015-03-04 12:36:25,677 - INFO - Starting new HTTPS connection (1): lb1.hubic.ovh.net
	[I 15-03-04 12:36:25] 192.168.1.13:50042-[] FTP session closed (disconnect).
	[1399] 2015-03-04 12:36:25,677 - INFO - 192.168.1.13:50042-[] FTP session closed (disconnect).
	[6780] 2015-03-04 12:36:25,695 - INFO - Starting new HTTPS connection (1): myServer.com
	/Library/Python/2.7/site-packages/requests-2.5.0-py2.7.egg/requests/packages/urllib3/connectionpool.py:734: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
	  InsecureRequestWarning)
	  […]

**Comment by reidrac** [comment]
Thanks for the info, I'll take a look when I have some free time!

**Comment by Tazintosh** [comment]
Thanks,

I also have to kind of behaviors:
• Either it will fail after reaching 512MB and stop,
• Either it will continue uploading as if everything was perfect, and give you a copy error message when done. (Imagine you're uploading dozen of GB, and you discover only at the end, less than 10% —true— has been done). During all the time, the line was actually —really— uploading those GB.

**Comment by Tazintosh** [comment]
Hey Juan,

I've some news:
Yesterday, I was really pissed off trying for weeks now to backup my contents and having so many problems while uploading — I'm not pointing any soft as we have yet no clue from where the problem can come, but I think you get my point, it's really frustrating :)
The 1/10 working was already a great statistic compared to the reality of the situation.

Then, —note that you'll find this obvious, as I did—, I focused and figured out that all problems where triggered in relation to the split. No matter my setting was 100MB (default value hubiC use as far as I know), or 512MB (my personal one) etc.
So I decided to set the split to 8192MB…

Yesterday late evening, I've started a backup that's more than 300GB (most probably all file > 1GB).
Guess what: not a single error at starting, then I go to bed and this morning, everything was already done for a while. Only two errors happened, and again, guess what: two files of 8.5 GB.
Means those two files are the only one that have triggered a split.

So here's my noob question: what's the split really for? Is what I'm doing "wrong" (but at least, it's working).

**Comment by reidrac** [comment]
swift has a default limit of 5GB per single object; you can't upload 8.5GB as a single object.

I'll review the code when I have the time (probably next week). There might be an issue that is more visible because the way hubic2swiftgate deals with expired tokens.

**Comment by reidrac** [comment]
Well, I can't see what's going on with this one. It's been a year, I'm closing it.

We we have more information, please reopen.
